### PR TITLE
XEP-0215 External Service Discovery

### PR DESCRIPTION
**Pull request by taiBsu**
PR check list:
- [x] Document your code
- [x] Add `\since QXmpp 1.X`, `QXMPP_EXPORT`
- [x] Fix doxygen warnings (see log when building with `-DBUILD_DOCUMENTATION=ON`)
- [x] Update `doc/doap.xml`
- [x] Add unit tests
- [x] Format the code: Run `clang-format -i src/<edited-file(s)> tests/<edited-file(s)>`

### CODE REVIEW

File: src/base/QXmppExternalServiceDiscoveryIq.cpp [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+            setAction(action.value());
+        }
+    }
+
+    if (attributes.contains("expires")) {
+        const QString &expiresAttribute { el.attribute("expires") };
+
+        QDateTime expires;
+        Qt::DateFormat dateFormat {
+            expiresAttribute.contains(QRegExp("\\.[0-9]{3}"))
```
Comment (JBBgameich) [inline src/base/QXmppExternalServiceDiscoveryIq.cpp:298]: QRegExp is replaced by QRegularExpression in Qt6. Would it be possible to use that instead?

Comment (taiBsu) [reply src/base/QXmppExternalServiceDiscoveryIq.cpp]: Fixed.

File: src/base/QXmppExternalServiceDiscoveryIq.cpp [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+    if (action().has_value()) {
+        const std::optional<Action> action { this->action() };
+
+        if (action.has_value()) {
+            helperToXmlAddAttribute(writer, "action", actionToString(action.value()));
+        }
+    }
+
+    if (expires().has_value()) {
+        helperToXmlAddAttribute(writer, "expires", expires().value().toString(Qt::ISODateWithMs));
```
Comment (lnjX) [inline src/base/QXmppExternalServiceDiscoveryIq.cpp:362]:
```suggestion
        helperToXmlAddAttribute(writer, "expires", d->expires->toString(Qt::ISODateWithMs));
```
`std::optional` supports the `->` operator :)

File: src/base/QXmppExternalServiceDiscoveryIq.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    ///
+    /// Describes the packet type of an external service IQ.
+    ///
+    /// \since QXmpp 1.6
+    ///
+    enum class Transport {
+        TCP,
+        UDP,
+        Other
```
[the remark this answers is not in the document]
Comment (taiBsu) [reply src/base/QXmppExternalServiceDiscoveryIq.h:39]: XEP-0215 says 

> transport  | The underlying transport protocol to be used when communicating with the service (typically either TCP or UDP).

which for me sounds like there were other possibilities as well which are not provided in the XEP, so I left an "Other" option here.

File: src/base/QXmppExternalServiceDiscoveryIq.cpp [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+{
+    QDomNamedNodeMap attributes { el.attributes() };
+
+    setHost(el.attribute("host"));
+    setType(el.attribute("type"));
+
+    d->action = actionFromString(el.attribute("action"));
+
+    if (attributes.contains("expires")) {
+        const QString &expiresAttribute { el.attribute("expires") };
```
Comment (lnjX) [inline src/base/QXmppExternalServiceDiscoveryIq.cpp:286]: `const &` of temporary

File: src/base/QXmppExternalServiceDiscoveryIq.cpp [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+
+        QDateTime expires;
+
+        expires = QXmppUtils::datetimeFromString(expiresAttribute);
+        expires.setTimeZone(QTimeZone::utc());
```
Comment (lnjX) [inline src/base/QXmppExternalServiceDiscoveryIq.cpp:291]: I don't think you need to set the time zone manually, or am I wrong?

File: src/base/QXmppExternalServiceDiscoveryIq.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    QXMPP_PRIVATE_DECLARE_RULE_OF_SIX(QXmppExternalService)
+
+    QString host() const;
+    void setHost(const QString &);
+
+    QString type() const;
+    void setType(const QString &);
+
+    std::optional<Action> action() const;
+    void setAction(const Action);
```
[the remark this answers is not in the document]
Comment (lnjX) [reply src/base/QXmppExternalServiceDiscoveryIq.h]: Ah sorry, can you do that for the other setters, too?

[the remark this answers is not in the document]
Comment (taiBsu) [reply src/base/QXmppExternalServiceDiscoveryIq.h]: Oops, my bad.

**Review by lnjX — APPROVED** [review]
:+1:

File: src/base/QXmppExternalServiceDiscoveryIq.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (melvo) [reply src/base/QXmppExternalServiceDiscoveryIq.h]: @taiBsu It is always about consistency :)

**Pull request merged**